### PR TITLE
[3.10] Fix Progressive caching to cache modules per custom menu assignment

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue #36102 and #36317 and #36323 .

### Summary of Changes
Same as #36193 for J 4.

Progressive caching caches the modules per position, however this breaks custom menu assignment for module.
I have add Itemid to cache key for Progressive caching, so it works correctly now.

upd: it was broken by  #35528

### Testing Instructions

Apply patch and please follow linked issue: #36102 and #36317 and #36323 

Create a custom module, and assign to one specific menu.
Enable Progressive caching, and visit that menu, then visit other pages.

### Actual result BEFORE applying this Pull Request
The module visible on all pages.

### Expected result AFTER applying this Pull Request
The module visible only on selected page.

### CODE REVIEW

**Comment by efrancis** [comment]
I have tested this item :white_check_mark: successfully on 60d445cb882736576ba2a0b085d561606c39a204<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/36324">issues.joomla.org/tracker/joomla-cms/36324</a>.</sub>

File: libraries/src/Document/HtmlDocument.php [hunk 1]
```diff
@@ -471,8 +472,17 @@ public function getBuffer($type = null, $name = null, $attribs = array())
...
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = CmsFactory::getCache('com_modules', '');
+			$hash = md5(
+				serialize(
+					array(
+						$name,
+						$attribs,
+						get_class($renderer),
```
Comment (PhilETaylor) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
						\get_class($renderer),
```

**Comment by richard67** [comment]
I've restored the previous human test result in the issue tracker since the commit which invalidated the test was just the change from `get_class(...)` to `\get_class(...)`, which doesn't change anything related to the test and it fine for me by review.

**Comment by wilsonge** [comment]
Adding a release blocker label here

File: libraries/src/Document/HtmlDocument.php [hunk 2]
```diff
@@ -11,6 +11,7 @@
 defined('JPATH_PLATFORM') or die;
 
 use Joomla\CMS\Cache\Cache;
+use Joomla\CMS\Factory as CmsFactory;
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]: Is there a specific reason to use the alias here? I saw it was used within J4 too but here it looks like its not needed right.

File: libraries/src/Document/HtmlDocument.php [hunk 3]
```diff
@@ -471,8 +472,17 @@ public function getBuffer($type = null, $name = null, $attribs = array())
...
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = CmsFactory::getCache('com_modules', '');
+			$hash = md5(
+				serialize(
+					array(
+						$name,
+						$attribs,
+						\get_class($renderer),
+						CmsFactory::getApplication()->input->get('Itemid')
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]: That is user provided content right? Can we please make sure we only go with an integer value here.

```suggestion
						CmsFactory::getApplication()->input->get('Itemid', 0, 'int')
```

Better would be to do an `$itemId = (int) CmsFactory::getApplication()->input->get('Itemid', 0, 'int');` above i guess.

**Review by zero-24 — COMMENTED** [review]
Hi @Fedik I'm sorry for the late reply here but I hope you can take a look here before the weekend so we can ship it with the next RC.

File: libraries/src/Document/HtmlDocument.php (hunk 2 above)
Comment (Fedik) [reply libraries/src/Document/HtmlDocument.php]: yea, it just a copy of joomla 4 ;)

File: libraries/src/Document/HtmlDocument.php (hunk 3 above)
Comment (Fedik) [reply libraries/src/Document/HtmlDocument.php]: okay, I will change

File: libraries/src/Document/HtmlDocument.php (hunk 2 above)
Comment (zero-24) [reply libraries/src/Document/HtmlDocument.php]: Ok than please use the native name here for now :)

**Comment by Fedik** [comment]
@zero-24 I have updated,
Drone cannot find mysql, but it something not related

**Comment by richard67** [comment]
@wilsonge As you had added the release blocker here, is the equivalent PR #36193 for 4.0 a release blocker, too? Up to now it doesn't have that label yet.

File: libraries/src/Document/HtmlDocument.php [hunk 4]
```diff
@@ -471,8 +472,17 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = Factory::getCache('com_modules', '');
+			$hash = md5(
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php:478]:
```suggestion
			$itemId = (int) Factory::getApplication()->input->get('Itemid', 0, 'int');
			$hash = md5(
```
```suggestion
			$hash = md5(
```

File: libraries/src/Document/HtmlDocument.php [hunk 5]
```diff
@@ -471,8 +472,17 @@ public function getBuffer($type = null, $name = null, $attribs = array())
...
+				serialize(
+					array(
+						$name,
+						$attribs,
+						\get_class($renderer),
+						(int) Factory::getApplication()->input->get('Itemid', 0, 'int')
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
						$itemId
```
```suggestion
						(int) Factory::getApplication()->input->get('Itemid', 0, 'int')
```

Comment (PhilETaylor) [reply libraries/src/Document/HtmlDocument.php]: `(int) Factory::getApplication()->input->get('Itemid', 0, 'int')` ????

isn't the first `(int)` redundant?

File: libraries/src/Document/HtmlDocument.php [hunk 6]
```diff
@@ -471,8 +472,18 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = Factory::getCache('com_modules', '');
+			$itemId = (int) Factory::getApplication()->input->get('Itemid', 0, 'int');
+			$hash = md5(
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php:478]:
```suggestion
			$hash =   md5(
```

File: libraries/src/Document/HtmlDocument.php [hunk 7]
```diff
@@ -471,8 +472,18 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = Factory::getCache('com_modules', '');
+			$itemId = (int) Factory::getApplication()->input->get('Itemid', 0, 'int');
+			$hash =   md5(
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
			$hash   = md5(
```

File: libraries/src/Document/HtmlDocument.php [hunk 8]
```diff
@@ -471,8 +472,18 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = Factory::getCache('com_modules', '');
+			$itemId = (int) Factory::getApplication()->input->get('Itemid', 0, 'int');
+			$hash   = md5(
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion

			$hash = md5(
```

File: libraries/src/Document/HtmlDocument.php [hunk 9]
```diff
@@ -471,8 +472,18 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache = Factory::getCache('com_modules', '');
```
Comment (zero-24) [inline libraries/src/Document/HtmlDocument.php]:
```suggestion
			$cache  = Factory::getCache('com_modules', '');
```

File: libraries/src/Document/HtmlDocument.php [hunk 10]
```diff
@@ -471,8 +472,19 @@ public function getBuffer($type = null, $name = null, $attribs = array())
 
 		if ($this->_caching == true && $type == 'modules' && $name !== 'debug')
 		{
-			$cache = \JFactory::getCache('com_modules', '');
-			$hash = md5(serialize(array($name, $attribs, null, get_class($renderer))));
+			$cache  = Factory::getCache('com_modules', '');
+			$itemId = (int) Factory::getApplication()->input->get('Itemid', 0, 'int');
```
Comment (PhilETaylor) [inline libraries/src/Document/HtmlDocument.php:476]:
```suggestion
			$itemId = Factory::getApplication()->input->get('Itemid', 0, 'int');
```

**Comment by zero-24** [comment]
I'm sorry for messing arround here again but I have hopefully improved the the codestyle now and moved the cast to integer out of the md5 call.

File: libraries/src/Document/HtmlDocument.php (hunk 10 above)
Comment (PhilETaylor) [reply libraries/src/Document/HtmlDocument.php:476]: `input->get('Itemid', 0, 'int')`should return an int right? so no need to type cast it further

Comment (zero-24) [reply libraries/src/Document/HtmlDocument.php:476]: Why? The cast has to happen else we could end up with an array of integers ;)

Comment (richard67) [reply libraries/src/Document/HtmlDocument.php:476]: Just wanted to type the same, can be an array sometimes.

Comment (PhilETaylor) [reply libraries/src/Document/HtmlDocument.php:476]: if you are getting an array from `input->get('Itemid', 0, 'int')` then something is seriously wrong...

Comment (PhilETaylor) [reply libraries/src/Document/HtmlDocument.php:476]: Also if it is an array, typecasting will give you an Itemid of 1 !!! https://3v4l.org/MhuFJ which is probably not what you want, especially as the default is set to `0` here...

Comment (Fedik) [reply libraries/src/Document/HtmlDocument.php:476]: It does not mater, it could stay default `cmd`, it used only for "cache" key

Comment (zero-24) [reply libraries/src/Document/HtmlDocument.php:476]: > Also if it is an array, typecasting will give you an Itemid of 1 !!! https://3v4l.org/MhuFJ which is probably not what you want, especially as the default is set to 0 here...

That would not break anything here as its just used as hashing key. ;)

> if you are getting an array from input->get('Itemid', 0, 'int') then something is seriously wrong...

I agree with you here but for now its the intended behavior (which i dont agree that this should be intended) of the filter lib and we can not change it for now for obvius reasons ;)

**Comment by zero-24** [comment]
Will merge it for now thanks guys :)

**Pull request merged**